sdk/python/kfp/compiler/compiler_test.py

```diff
@@ -3180,5 +3180,232 @@ def pipeline_component(input_list: Input[List[Artifact]]):
+            'comp-print-op')
+        self.assertEqual(
+            pipeline_spec.root.dag.tasks['print-op-4'].component_ref.name,
+            'comp-print-op')
+
+        # check task specific input/ output information is not lost
+        self.assertEqual(
+            pipeline_spec.root.dag.tasks['print-op-2'].inputs
+            .parameters['message'].runtime_value.constant.string_value,
+            'constant')
+        self.assertEqual(
+            pipeline_spec.root.dag.tasks['print-op-3'].inputs
+            .parameters['message'].component_input_parameter, 'pipeline_input')
+        self.assertEqual(
+            pipeline_spec.root.dag.tasks['print-op-4'].inputs
+            .parameters['message'].task_output_parameter.producer_task,
+            'identity')
+        self.assertEqual(
+            pipeline_spec.root.dag.tasks['print-op-4'].inputs
+            .parameters['message'].task_output_parameter.output_parameter_key,
+            'Output')
+
+    def test_task_specific_executor_case_would_not_dedupe(self):
+
+        @dsl.component
+        def print_op(message: str = 'default'):
+            print(message)
+
+        @dsl.pipeline()
+        def my_pipeline():
+            task1 = print_op()
+            task2 = print_op().set_cpu_limit('4').set_memory_limit('14Gi')
+
+        pipeline_spec = my_pipeline.pipeline_spec
+
+        self.assertEqual(len(pipeline_spec.components), 2)
+
+        # check both tasks component spec and executors exist
+        self.assertIn('comp-print-op', pipeline_spec.components)
+        self.assertIn(
+            'exec-print-op', pipeline_spec.deployment_spec.fields['executors']
+            .struct_value.fields)
+        self.assertIn('comp-print-op-2', pipeline_spec.components)
+        self.assertIn(
+            'exec-print-op-2', pipeline_spec.deployment_spec.fields['executors']
+            .struct_value.fields)
+
+        # check both component specs contains the correct executor label
+        self.assertEqual(
+            pipeline_spec.components['comp-print-op'].executor_label,
+            'exec-print-op')
+        self.assertEqual(
+            pipeline_spec.components['comp-print-op-2'].executor_label,
+            'exec-print-op-2')
+
+        # check tasks point to corrrect component spec reference
+        self.assertEqual(
+            pipeline_spec.root.dag.tasks['print-op'].component_ref.name,
+            'comp-print-op')
+        self.assertEqual(
+            pipeline_spec.root.dag.tasks['print-op-2'].component_ref.name,
+            'comp-print-op-2')
+
+        # check information of resource spec is not lost
+        self.assertEqual(
+            4.0, pipeline_spec.deployment_spec.fields['executors'].struct_value
+            .fields['exec-print-op-2'].struct_value['container']
+            ['resources'].fields['cpuLimit'].number_value)
+        self.assertEqual(
+            15.032385536, pipeline_spec.deployment_spec.fields['executors']
```

Comment:
For float comparison, I think we should use [`assertAlmostEqual`](`https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertAlmostEqual`).